perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -236,11 +245,28 @@ public static boolean isAuthorizedForAttribute(PerunSession sess, ActionType act
 					}
 				}
 			}
-//			if (roles.contains(Role.SELF)) ; //Not Allowed
+//			if (roles.containsKey(Role.SELF)) ; //Not Allowed
 		} else if (user != null && facility != null) {
-			if (roles.contains(Role.FACILITYADMIN)) if (isAuthorized(sess, Role.FACILITYADMIN, facility)) return true;
-			if (roles.contains(Role.SELF)) if (isAuthorized(sess, Role.SELF, user)) return true;
-			if (roles.contains(Role.VOADMIN)) {
+			if (roles.containsKey(Role.FACILITYADMIN)) if (isAuthorized(sess, Role.FACILITYADMIN, facility)) return true;
+			if (roles.containsKey(Role.SELF)) {
+				if (roles.get(Role.SELF).contains(ActionType.READ_PUBLIC) || roles.get(Role.SELF).contains(ActionType.WRITE_PUBLIC)) return true;
+				if (roles.get(Role.SELF).contains(ActionType.READ) || roles.get(Role.SELF).contains(ActionType.WRITE)) {
+					if (isAuthorized(sess, Role.SELF, user)) return true;
+				}
+				if (roles.get(Role.SELF).contains(ActionType.READ_VO) || roles.get(Role.SELF).contains(ActionType.WRITE_VO)) {
+					List<Vo> userVos = getPerunBl().getUsersManagerBl().getVosWhereUserIsMember(sess, user);
+					List<Resource> resourcesFromFacility = getPerunBl().getFacilitiesManagerBl().getAssignedResources(sess, facility);
+					for (Vo voElement : userVos) {
+						List<Resource> resourcesFromVo = getPerunBl().getResourcesManagerBl().getResources(sess, voElement);
+						for (Resource resourceElement : resourcesFromVo) {
+							if (resourcesFromFacility.contains(resourceElement)) {
+								return true;
```

Comment:
This is similar case as the previous one. This first part verifies the relation of objects from the attribute.

But you also need to verify that the current principal is in the same Vo like the user from the attribute.